omc3/optics_measurements/crdt.py

```diff
@@ -0,0 +1,228 @@
+    {'order':"Octupole", 'term': "F_NO4", 'func': Aover8BC, 'lines': {'A': 'X12', 'B': 'X10', 'C': 'Y01'}, 'sign':-1},
+    {'order':"Octupole", 'term': "F_NO3", 'func': Aover8B, 'lines': {'A': 'X30', 'B': 'X10'}, 'sign':1},
+    {'order':"Octupole", 'term': "F_NO2", 'func': Aover8BC, 'lines': {'A': 'X_12', 'B': 'X10', 'C': 'Y01'}, 'sign':1},
+    {'order':"Octupole", 'term': "F_NO1", 'func': Aover8BC, 'lines': {'A': 'Y2_1', 'B': 'Y01', 'C': 'X10'}, 'sign':1},
+    {'order':"Octupole", 'term': "F_NO0", 'func': Aover8BC, 'lines': {'A': 'Y21', 'B': 'Y01', 'C': 'X10'}, 'sign':-1},
+]
+
+
+def calculate(measure_input, input_files, header):
+    """
+
+    Args:
+        measure_input:
+        input_files:
+        header:
+    Returns:
+
+    """
+    LOGGER.info("Start of CRDT analysis")
+    for plane in PLANES:
+        input_files[plane] = unscale_amps(input_files[plane], plane)
+        input_files[plane] = scale_amps_with_sqrtbeta(input_files[plane], measure_input, plane)
+
+    joined_dfs = joined_planes(input_files)
+
+    for crdt in CRDTS:
+        LOGGER.debug(f"Processing CRDT {crdt['term']}")
+        result_dfs = []
+        for joined_df in joined_dfs:
+            result_dfs.append(process_crdt(joined_df, crdt))
+        df = average_results(result_dfs=result_dfs,
+                             union_columns=['NAME', 'S'],
+                             merge_columns=[AMPLITUDE, f'{ERR}{AMPLITUDE}', PHASE, f'{ERR}{PHASE}'],
+                             merge_functions=[stats.weighted_nanmean, stats.weighted_nanrms,
+                                              stats.circular_nanmean, stats.circular_nanerror],
+                             index="NAME")
+        write(df, header, measure_input, crdt['order'], crdt['term'])
+
+
+def scale_amps_with_sqrtbeta(input_files, meas_input, plane):
+    model_beta = meas_input.accelerator.model[f'BET{plane}']
+    processed_files = []
+    for lin_df in input_files:
+        cols = [col for col in lin_df.columns.to_numpy() if col.startswith('AMP')]
+        lin_df.loc[:, cols] = lin_df.loc[:, cols].div(model_beta**0.5, axis="index")
+        processed_files.append(lin_df)
+    return processed_files
+
+
+def unscale_amps(input_files, plane):
+    processed_files = []
+    for lin_df in input_files:
+        cols = [col for col in lin_df.columns.to_numpy() if col.startswith('AMP')]
+        cols.remove(f"AMP{plane}")
+        lin_df[f"AMP{plane}"] = lin_df.loc[:, f"AMP{plane}"].to_numpy()/2.
+        lin_df.loc[:, cols] = lin_df.loc[:, cols].mul(lin_df.loc[:, f'AMP{plane}'], axis="index")
+        processed_files.append(lin_df)
+    return processed_files
+
+def process_crdt(joined_df, crdt):
+    df = pd.DataFrame(index=joined_df.index, data={'S': joined_df['S']})
+
+    lines = {}
+    phases = {}
+    errlines = {}
+    errphases = {}
+    for data_dict, prefix in zip([lines, phases, errlines, errphases], ['AMP', 'FREQ', 'ERRAMP', 'ERRFREQ']):
```

Comment:
Fom constants?

Reply:
fixed, but used naming convention for crdt only, so no global constant